core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -123,33 +130,35 @@ DeleteFile[] forEntry(ManifestEntry<DataFile> entry) {
   }
 
   DeleteFile[] forDataFile(long sequenceNumber, DataFile file) {
+    if (isEmpty) {
+      return NO_DELETES;
+    }
+
     Pair<Integer, StructLikeWrapper> partition = partition(file.specId(), file.partition());
-    Pair<long[], DeleteFile[]> partitionDeletes = sortedDeletesByPartition.get(partition);
+    DeleteFileGroup partitionDeletes = deletesByPartition.get(partition);
+
+    if (globalDeletes == null && partitionDeletes == null) {
+      return NO_DELETES;
+    }
 
-    Stream<DeleteFile> matchingDeletes;
+    Stream<IndexedDeleteFile> matchingDeletes;
     if (partitionDeletes == null) {
-      matchingDeletes = limitBySequenceNumber(sequenceNumber, globalSeqs, globalDeletes);
+      matchingDeletes = globalDeletes.limit(sequenceNumber);
     } else if (globalDeletes == null) {
-      matchingDeletes =
-          limitBySequenceNumber(
-              sequenceNumber, partitionDeletes.first(), partitionDeletes.second());
+      matchingDeletes = partitionDeletes.limit(sequenceNumber);
     } else {
-      matchingDeletes =
-          Stream.concat(
-              limitBySequenceNumber(sequenceNumber, globalSeqs, globalDeletes),
-              limitBySequenceNumber(
-                  sequenceNumber, partitionDeletes.first(), partitionDeletes.second()));
+      Stream<IndexedDeleteFile> matchingGlobalDeletes = globalDeletes.limit(sequenceNumber);
+      Stream<IndexedDeleteFile> matchingPartitionDeletes = partitionDeletes.limit(sequenceNumber);
+      matchingDeletes = Stream.concat(matchingGlobalDeletes, matchingPartitionDeletes);
     }
 
     return matchingDeletes
-        .filter(
-            deleteFile ->
-                canContainDeletesForFile(file, deleteFile, specsById.get(file.specId()).schema()))
+        .filter(deleteFile -> canApply(file, deleteFile, specsById.get(file.specId()).schema()))
+        .map(IndexedDeleteFile::wrapped)
         .toArray(DeleteFile[]::new);
   }
 
-  private static boolean canContainDeletesForFile(
-      DataFile dataFile, DeleteFile deleteFile, Schema schema) {
+  private static boolean canApply(DataFile dataFile, IndexedDeleteFile deleteFile, Schema schema) {
```

Comment:
Renamed for better formatting of the closure above.

Reply:
This seems to lose a lot of the value of the former name. `canContainDeletesForFile` is clear, but `canApply` doesn't say the same thing. Is it "should" apply? It also omits the criteria. Not a big deal since this is internal, but I'd prefer the original name and just let the closure be a bit messier.

Reply:
I'll play around.

Reply:
I reverted the name.